File: go/euler02.go
// euler02 находит сумму всех четных элементов ряда Фибоначчи,
// которые не превышают четыре миллиона.
//
// Решение задачи Эйлера №2.
package main

import (
	"fmt"
	"time"
)

// fibonacci возвращает канал с числами Фибоначчи.
func fibonacci() chan int {
	ch := make(chan int)
	go func() {
		f1, f2 := 0, 1
		for {
			ch <- f1
			f1, f2 = f2, f1+f2
		}
	}()
	return ch
}

// euler02 решает задачу Эйлера №2.
//
// Находит сумму всех четных элементов ряда Фибоначчи,
// которые не превышают четыре миллиона.
func euler02() {
	n := 4_000_000

	// Вариант 1
	startTime := time.Now()
	oddFbSum := 0
	iterator := fibonacci()
	for fbNum := range iterator {
		if fbNum >= n {
			break
		}
		if fbNum%2 == 0 {
			oddFbSum += fbNum
		}
	}
	fmt.Println(oddFbSum)
	fmt.Println(time.Since(startTime))

	// Вариант 2
	startTime = time.Now()
	sum := 0
	iterator = fibonacci()
	for fbNum := range iterator {
		if fbNum >= n {
			break
		}
		if fbNum%2 == 0 {
			sum += fbNum
		}
	}
	fmt.Println(sum)
	fmt.Println(time.Since(startTime))
}

func main() {
	euler02()
}

File: python/euler02.py
# Найдите сумму всех четных элементов ряда Фибоначчи, которые не превышают
# четыре миллиона.

from datetime import datetime
from functools import reduce
from itertools import takewhile
from typing import Generator


def fibonacci(f1: int = 0, f2: int = 1) -> Generator[int, None, None]:
    """
    Итерируемая функция вычисления следующего числа Фибоначчи.

    Args:
        f1: Первое число последовательности.
        f2: Второе число последовательности.

    Yields:
        Очередное число Фибоначчи.
    """
    while True:
        yield f1
        f1, f2 = f2, f1 + f2


def euler02() -> None:
    """
    Решение задачи Эйлера №2.

    Найдите сумму всех четных элементов ряда Фибоначчи,
    которые не превышают четыре миллиона.
    """
    n = 4_000_000

    # 1 Вариант
    start_time = datetime.now()
    odd_fb_sum, iterator = 0, fibonacci()
    while (fb_num := next(iterator)) < n:
        odd_fb_sum += fb_num if fb_num % 2 == 0 else 0
    print(odd_fb_sum)
    print(datetime.now() - start_time)

    # 2 Вариант
    start_time = datetime.now()
    print(reduce(
        lambda x, y: x + y,
        filter(lambda s: s % 2 == 0, takewhile(lambda f: f < n, fibonacci()))
    ))
    print(datetime.now() - start_time)


if __name__ == "__main__":
    euler02()
